### PR TITLE
SF-1479: Add job history and audit commands.

### PR DESCRIPTION
**Pull request by mikesae**

### CODE REVIEW

File: src/commands/jobs/audit.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import {auditOutputFormat} from './util';
+
+export = {
+	command: 'audit [id]',
+	desc: 'Show job execution audit',
+	builder: yargs => {
+		yargs.options({
+			'execution': {
+				default: undefined,
+				description: 'An execution index',
```
Comment (smbecker) [inline src/commands/jobs/audit.ts]: `An execution ID`

File: src/commands/jobs/history.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import {historyOutputFormat} from './util';
+
+export = {
+	command: 'history [id]',
+	desc: 'Show job execution history',
+	builder: yargs => {
+		yargs.options({
+			'execution': {
+				default: undefined,
+				description: 'An execution index, last or current',
```
Comment (smbecker) [inline src/commands/jobs/history.ts]: `An execution ID, last or current`.

File: src/commands/jobs/util.ts [hunk 3]
```diff
@@ -57,4 +57,83 @@ const detailOutputFormat = {
...
+			property: 'execution_id'
+		},
+		{
+			header: 'FromSource',
+			property: 'from_source'
+		},
+		{
+			header: 'ToDestination',
+			property: 'to_destination',
+			transform: value => new Date(value)
```
Comment (smbecker) [inline src/commands/jobs/util.ts]: This `transform` doesn't seem right. `to_destination` is a boolean field.

File: src/commands/jobs/util.ts [hunk 4]
```diff
@@ -57,4 +57,83 @@ const detailOutputFormat = {
...
+		},
+		{	
+			header: 'Level',
+			property: 'level'
+		},
+		{	
+			header: 'Event',
+			property: 'event'
+		}
```
Comment (smbecker) [inline src/commands/jobs/util.ts]: Should `message` be included?

File: src/sdk/http/http-client.ts [hunk 5]
```diff
@@ -1,10 +1,9 @@
 import { URL } from 'url';
-import { Readable } from 'stream';
 
 const API_VERSION = 'v1';
 
 export interface IHttpClient {
-	get(path: string, params?: any): Promise<any>;
+	get(path: string, params?: any, csvResponse?: boolean): Promise<any>;
```
Comment (smbecker) [inline src/sdk/http/http-client.ts]: I would remove this parameter and update below to only parse the JSON if `content-type: application/json`.

File: src/sdk/models/jobs.ts [hunk 6]
```diff
@@ -43,3 +43,49 @@ export interface TransferOptions {
...
+	units?: string;
+}
+
+export interface JobExecutionStatistics {
+	bytes?: number;
+	files?: number;
+	files_pending?: number;
+	rate_limits?: number;
+	versions?: number;
+	versions_pending?: number;
```
Comment (smbecker) [inline src/sdk/models/jobs.ts]: Add `[name: string]: number` to allow access to other stats not exposed here.

File: src/util/formatter.ts [hunk 7]
```diff
@@ -45,6 +45,16 @@ export class OutputFormatter {
...
 		}
 	}
+	
+	writeWarning(message: string, force: boolean = false): void {
+		if (!this.outputJson || force) {
+			console.warn((message as any).yellow);
+		}
+	}
+	
+	writeCsv(output: string): void {
```
Comment (smbecker) [inline src/util/formatter.ts]: `writeText`

File: src/commands/jobs/audit.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+					result = await client.jobs.getAuditCsvList(argv.id, {include: ['all']});
+				}
+				output.writeCsv(result);
+			} else {
+				if (argv.execution) {
+					result = await client.jobs.getAudit(argv.id, argv.execution, {include: ['all']});
+				} else {
+					result = await client.jobs.getAuditList(argv.id, {include: ['all']});
+				}
+				output.writeItem(result, auditOutputFormat);
```
Comment (smbecker) [inline src/commands/jobs/audit.ts]: Also, shouldn't this be `writeTable` if using writing the list of audit records?

File: src/sdk/models/jobs.ts (hunk 6 above)
Comment (mikesae) [reply src/sdk/models/jobs.ts]: That's a new one on me! https://devdocs.io/typescript/handbook/interfaces#optional-properties

File: src/commands/jobs/audit.ts (hunk 8 above)
Comment (mikesae) [reply src/commands/jobs/audit.ts]: Yes, same for history cmd too.

File: src/commands/jobs/util.ts (hunk 4 above)
Comment (mikesae) [reply src/commands/jobs/util.ts]: Do you mean `reason` ?  If so, I will add.

**Comment by mikesae** [comment]
Still a few test failures.  Working on it now.

File: src/sdk/http/http-client.ts [hunk 9]
```diff
@@ -153,7 +153,7 @@ export abstract class HttpClient<TRequest, TResponse> implements IHttpClient {
 		});
 	}
 
-	get(path: string, params?: any): Promise<any> {
+	get(path: string, params?: any, csvResponse?: boolean): Promise<any> {
```
Comment (smbecker) [inline src/sdk/http/http-client.ts]: Missed removing this parameter.

**Pull request merged**